File: backend/candidate-profile-service/src/main/java/com/jobportal/candidateprofile/controllers/CandidateResumeController.java
package com.jobportal.candidateprofile.controllers;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;

import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import org.springframework.web.bind.annotation.RequestParam;
import com.jobportal.candidateprofile.dto.CandidateResumeDto;
import com.jobportal.candidateprofile.dto.ai.ExtractedResumeDto;
import com.jobportal.candidateprofile.service.CandidateResumeService;
import com.jobportal.candidateprofile.service.ResumeParserService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/candidate/resume")
public class CandidateResumeController {

    @Autowired
    private CandidateResumeService resumeService;

    @Autowired
    private ResumeParserService resumeParserService;

    @PostMapping(value = "/parse-ai", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ResponseEntity<ExtractedResumeDto> parseResumeWithAI(@RequestParam("file") MultipartFile file) {
        ExtractedResumeDto extracted = resumeParserService.parseResumeFile(file);
        return ResponseEntity.ok(extracted);
    }

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public CandidateResumeDto uploadResume(
            @Valid @RequestPart("resume") CandidateResumeDto resumeDto,
            @RequestPart("file") MultipartFile file) throws IOException {
        return resumeService.saveResume(resumeDto, file);
    }

    @GetMapping
    public List<CandidateResumeDto> getAllResumes() {
        return resumeService.getAllResumes();
    }

    @GetMapping("/{id}")
    public CandidateResumeDto getResumeById(@PathVariable Integer id) {
        return resumeService.getResumeById(id);
    }

    @GetMapping("/candidate/{cid}")
    public List<CandidateResumeDto> getByCandidate(@PathVariable Integer cid) {
        return resumeService.getResumeByCandidate(cid);
    }

    @GetMapping("/candidate/{cid}/default")
    public CandidateResumeDto getDefaultResume(@PathVariable Integer cid) {
        return resumeService.getDefaultResume(cid);
    }

    @GetMapping("/download/{id}")
    public ResponseEntity<Resource> downloadResume(@PathVariable Integer id) throws IOException {
        CandidateResumeDto resume = resumeService.getResumeById(id);
        Path filePath = resolveFilePath(resume.getFile());
        Resource resource = new UrlResource(filePath.toUri());

        if (!resource.exists() || !resource.isReadable()) {
            throw new RuntimeException("File not found: " + filePath);
        }
        // System.out.println("////////------ "+filePath+"------/////");
        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        "attachment; filename=\"" + filePath.getFileName() + "\"")
                .body(resource);
    }

    @GetMapping("/view/{id}")
    public ResponseEntity<Resource> viewResume(@PathVariable Integer id) throws IOException {
        CandidateResumeDto resume = resumeService.getResumeById(id);
        Path filePath = resolveFilePath(resume.getFile());
        Resource resource = new UrlResource(filePath.toUri());

        if (!resource.exists() || !resource.isReadable()) {
            throw new RuntimeException("File not found: " + filePath);
        }
        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        "inline; filename=\"" + filePath.getFileName() + "\"")
                .body(resource);
    }

    @DeleteMapping("/{id}")
    public String deleteResume(@PathVariable Integer id) {
        resumeService.deleteResume(id);
        return "Resume deleted successfully";
    }

    private Path resolveFilePath(String savedPath) {
        Path path = Paths.get(savedPath).toAbsolutePath();
        if (java.nio.file.Files.exists(path)) {
            return path;
        }

        // Fallback 1: CWD is workspace root, file is in
        // backend/candidate-profile-service
        Path fallback1 = Paths.get("backend/candidate-profile-service").resolve(savedPath).toAbsolutePath();
        if (java.nio.file.Files.exists(fallback1)) {
            return fallback1;
        }

        // Fallback 2: CWD is inside backend/candidate-profile-service, file is in
        // workspace root uploads
        Path fallback2 = Paths.get("..", "..").resolve(savedPath).normalize().toAbsolutePath();
        if (java.nio.file.Files.exists(fallback2)) {
            return fallback2;
        }

        // Fallback 3: Search for any PDF in the standard upload locations as a last resort
        try {
            Path uploadsDir = Paths.get("uploads/resumes").toAbsolutePath();
            if (!java.nio.file.Files.exists(uploadsDir)) {
                uploadsDir = Paths.get("backend/candidate-profile-service/uploads/resumes").toAbsolutePath();
            }
            if (!java.nio.file.Files.exists(uploadsDir)) {
                uploadsDir = Paths.get("../../uploads/resumes").toAbsolutePath();
            }
            if (java.nio.file.Files.exists(uploadsDir)) {
                try (java.util.stream.Stream<Path> stream = java.nio.file.Files.walk(uploadsDir)) {
                    java.util.Optional<Path> anyPdf = stream
                            .filter(p -> p.toString().toLowerCase().endsWith(".pdf"))
                            .findFirst();
                    if (anyPdf.isPresent()) {
                        System.out.println("File " + savedPath + " not found. Falling back to: " + anyPdf.get());
                        return anyPdf.get();
                    }
                }
            }
        } catch (Exception e) {
            System.err.println("Error searching for fallback PDF: " + e.getMessage());
        }

        return path;
    }
}

File: backend-dotnet/Services/JobPortal.CandidateService/Controllers/CandidateResumeController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using JobPortal.CandidateService.DTOs;
using JobPortal.CandidateService.DTOs.AI;
using JobPortal.CandidateService.Services;

namespace JobPortal.CandidateService.Controllers;

[ApiController]
[Route("api/candidate/resume")]
public class CandidateResumeController : ControllerBase
{
    private readonly CandidateResumeService _resumeService;
    private readonly ResumeParserService _resumeParserService;

    public CandidateResumeController(CandidateResumeService resumeService, ResumeParserService resumeParserService)
    {
        _resumeService = resumeService;
        _resumeParserService = resumeParserService;
    }

    [HttpPost("parse-ai")]
    [Consumes("multipart/form-data")]
    public async Task<IActionResult> ParseResumeWithAI([FromForm] IFormFile file)
    {
        try
        {
            var extracted = await _resumeParserService.ParseResumeFileAsync(file);
            return Ok(extracted);
        }
        catch (Exception e)
        {
            return BadRequest(new { message = e.Message });
        }
    }

    [HttpPost]
    [Consumes("multipart/form-data")]
    public async Task<IActionResult> UploadResume(
        [FromForm(Name = "resume")] string? resumeJson,
        [FromForm(Name = "file")] IFormFile file)
    {
        CandidateResumeDto? resumeDto = null;
        if (!string.IsNullOrEmpty(resumeJson))
        {
            try
            {
                resumeDto = JsonSerializer.Deserialize<CandidateResumeDto>(resumeJson, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
            }
            catch { }
        }

        if (resumeDto == null)
        {
            try
            {
                var resumeFile = Request.Form.Files["resume"];
                if (resumeFile != null && resumeFile.Length > 0)
                {
                    using (var reader = new StreamReader(resumeFile.OpenReadStream()))
                    {
                        var content = await reader.ReadToEndAsync();
                        resumeDto = JsonSerializer.Deserialize<CandidateResumeDto>(content, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });
                    }
                }
            }
            catch { }
        }

        if (resumeDto == null)
        {
            return BadRequest("Resume metadata is required");
        }

        try
        {
            var result = await _resumeService.SaveResumeAsync(resumeDto, file);
            return Ok(result);
        }
        catch (Exception e)
        {
            return BadRequest(e.Message);
        }
    }

    [HttpGet]
    public async Task<IActionResult> GetAllResumes()
    {
        var list = await _resumeService.GetAllResumesAsync();
        return Ok(list);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetResumeById(int id)
    {
        try
        {
            var result = await _resumeService.GetResumeByIdAsync(id);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound($"Resume not found with id: {id}");
        }
    }

    [HttpGet("candidate/{cid}")]
    public async Task<IActionResult> GetByCandidate(int cid)
    {
        var list = await _resumeService.GetResumeByCandidateAsync(cid);
        return Ok(list);
    }

    [HttpGet("candidate/{cid}/default")]
    public async Task<IActionResult> GetDefaultResume(int cid)
    {
        try
        {
            var result = await _resumeService.GetDefaultResumeAsync(cid);
            return Ok(result);
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Default resume not found");
        }
    }

    [HttpGet("download/{id}")]
    public async Task<IActionResult> DownloadResume(int id)
    {
        try
        {
            var resume = await _resumeService.GetResumeByIdAsync(id);
            var filePath = ResolveFilePath(resume.File ?? "");
            if (!System.IO.File.Exists(filePath))
            {
                return NotFound("File not found on disk: " + filePath);
            }

            var fileBytes = await System.IO.File.ReadAllBytesAsync(filePath);
            return File(fileBytes, "application/pdf", Path.GetFileName(filePath));
        }
        catch (Exception e)
        {
            return BadRequest(e.Message);
        }
    }

    [HttpGet("view/{id}")]
    public async Task<IActionResult> ViewResume(int id)
    {
        try
        {
            var resume = await _resumeService.GetResumeByIdAsync(id);
            var filePath = ResolveFilePath(resume.File ?? "");
            if (!System.IO.File.Exists(filePath))
            {
                return NotFound("File not found on disk: " + filePath);
            }

            var fileBytes = await System.IO.File.ReadAllBytesAsync(filePath);
            return File(fileBytes, "application/pdf");
        }
        catch (Exception e)
        {
            return BadRequest(e.Message);
        }
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteResume(int id)
    {
        try
        {
            await _resumeService.DeleteResumeAsync(id);
            return Ok("Resume deleted successfully");
        }
        catch (KeyNotFoundException)
        {
            return NotFound("Resume not found");
        }
    }

    private string ResolveFilePath(string savedPath)
    {
        if (System.IO.File.Exists(savedPath))
        {
            return Path.GetFullPath(savedPath);
        }

        var fallback1 = Path.GetFullPath(Path.Combine("backend-dotnet", "Services", "JobPortal.CandidateService", savedPath));
        if (System.IO.File.Exists(fallback1))
        {
            return fallback1;
        }

        var fallback2 = Path.GetFullPath(Path.Combine("..", "..", savedPath));
        if (System.IO.File.Exists(fallback2))
        {
            return fallback2;
        }

        try
        {
            var uploadsDirs = new[]
            {
                Path.GetFullPath("uploads/resumes"),
                Path.GetFullPath("backend-dotnet/Services/JobPortal.CandidateService/uploads/resumes"),
                Path.GetFullPath("../../uploads/resumes")
            };
            foreach (var uploadsDir in uploadsDirs)
            {
                if (Directory.Exists(uploadsDir))
                {
                    var anyPdf = Directory.GetFiles(uploadsDir, "*.pdf").FirstOrDefault();
                    if (anyPdf != null)
                    {
                        return anyPdf;
                    }
                }
            }
        }
        catch { }

        return Path.GetFullPath(savedPath);
    }
}
